src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs

```diff
@@ -140,48 +157,14 @@ public async Task<bool> IsFullyLoadedAsync(CancellationToken cancellationToken)
                 return !status.IsInProgress;
             }
 
-            private async Task<IVsOperationProgressStageStatusForSolutionLoad> GetProgressStageStatusAsync(CancellationToken cancellationToken)
-            {
-                var service = await _serviceProvider.GetServiceAsync<SVsOperationProgress, IVsOperationProgressStatusService>(throwOnFailure: false)
-                                                    .WithCancellation(cancellationToken).ConfigureAwait(false);
-
-                return service?.GetStageStatusForSolutionLoad(CommonOperationProgressStageIds.Intellisense);
-            }
-
-            private async Task EnsureInitializationAsync(CancellationToken cancellationToken)
+            private async ValueTask<IVsOperationProgressStageStatusForSolutionLoad?> GetProgressStageStatusAsync(CancellationToken cancellationToken)
             {
-                using (await _initializationGate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+                if (_progressStageStatus.Task.IsCompleted)
                 {
-                    if (_initialized)
-                    {
-                        return;
-                    }
-
-                    _initialized = true;
-
-                    _loadHubClientPackage = _threadingContext.JoinableTaskFactory.RunAsync(async () =>
-                    {
-                        // Use the disposal token, since the caller's cancellation token will apply instead to the
-                        // JoinAsync operation.
-                        await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(_threadingContext.DisposalToken);
-
-                        // Make sure the HubClient package is loaded, since we rely on it for proffered OOP services
-                        var shell = await _serviceProvider.GetServiceAsync<SVsShell, IVsShell7>().ConfigureAwait(true);
-                        Assumes.Present(shell);
-
-                        await shell.LoadPackageAsync(Guids.GlobalHubClientPackageGuid);
-                    });
-
-                    // with IAsyncServiceProvider, to get a service from BG, there is not much else
-                    // we can do to avoid this pattern to subscribe to events
-                    var status = await GetProgressStageStatusAsync(cancellationToken).ConfigureAwait(false);
-                    if (status == null)
-                    {
-                        return;
-                    }
-
-                    status.PropertyChanged += (_, e) => this.StatusChanged?.Invoke(this, EventArgs.Empty);
+                    return await _progressStageStatus.Task.ConfigureAwait(false);
                 }
+
+                return await _progressStageStatus.JoinAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
```suggestion
                    return await _progressStageStatus.Task.ConfigureAwait(false);
                }

                return await _progressStageStatus.JoinAsync(cancellationToken).ConfigureAwait(false);
```

i don't actually understand this bit...   why not just always do: 

```return await _progressStageStatus.Task.ConfigureAwait(false);```

? 

If tehre is substantive difference, please explain it with a comment in the code as hte pattern is not self evident to me.  Thanks!

Reply:
> i don't actually understand this bit... why not just always do:
>
> `return await _progressStageStatus.Task.ConfigureAwait(false);`

This is never valid (violation of the third threading rule):

https://github.com/microsoft/vs-threading/blob/master/doc/threading_rules.md#rule-3-if-ever-awaiting-work-that-was-started-earlier-that-work-must-be-joined

It is likely that this optimization path can be removed once vs-threading updates `JoinAsync` to have a proper fast path for completed tasks.

Reply:
i don't actually see how it's a violation of the third rule.  it seems to also say that it's fine to do `await longRunningAsyncWork;`  So i would expect to just see that here.

Reply:
➡️ Added a comment clarifying that the strange code paths are a workaround for a specific limitation which should be addressed in the future.

Reply:
🔗 Also filed https://github.com/microsoft/vs-threading/issues/697 to make the threading rules more clear about this syntax.